### PR TITLE
pkg/ns: introduce error types indicate NS verification

### PR DESCRIPTION
**Pull request by steveeJ**
Depends on #230.

### CODE REVIEW

File: pkg/ns/ns.go [hunk 1]
```diff
@@ -81,11 +81,23 @@ const (
...
+func (e NSPathNotNSErr) Error() string { return e.msg }
+
+func IsNSorErr(nspath string) error {
 	stat := syscall.Statfs_t{}
-	if err = syscall.Statfs(nspath, &stat); err != nil {
-		err = fmt.Errorf("failed to Statfs %s: %v", nspath, err)
-		return
+	if err := syscall.Statfs(nspath, &stat); err != nil {
+		if os.IsNotExist(err) {
+			err = NSPathNotExistErr{msg: fmt.Sprintf("failed to Statfs %s: %v", nspath, err)}
```
Comment (iaguis) [inline pkg/ns/ns.go]: `%q` on nspath?

File: pkg/ns/ns.go [hunk 2]
```diff
@@ -81,11 +81,23 @@ const (
...
+		} else {
+			err = fmt.Errorf("failed to Statfs %s: %v", nspath, err)
```
Comment (iaguis) [inline pkg/ns/ns.go]: ditto

File: pkg/ns/ns.go [hunk 3]
```diff
@@ -95,33 +107,29 @@ func IsNS(nspath string) (isNS bool, msg string, err error) {
...
 		return nil, err
 	}
-	if !isNS {
-		return nil, fmt.Errorf("no network namespace detected on %s: %s", nspath, msg)
-	}
 
 	fd, err := os.Open(nspath)
 	if err != nil {
-		return nil, fmt.Errorf("Failed to open %v: %v", nspath, err)
+		return nil, err
```
Comment (dcbw) [inline pkg/ns/ns.go]: for consistency, should this check os.IsNotExist() and return NSPathNotExistErr?  The path already got verified by IsNSorErr() but there's a very small window between that and the Open() where if could get deleted.

Comment (steveeJ) [inline pkg/ns/ns.go]: Alternatively I could move the call to `Open()` to the beginning of the function. Little chance it gets created and unmounted in the meantime ;-)

Comment (steveeJ) [inline pkg/ns/ns.go]: On second thought, this is not really something I'd like to consider. There is also a small window between the call to `IsNSorErr()` and the actual return of that error value, but yet it doesn't make sense to check it again. We're talking about very short periods of time, when the error we're probably trying to catch is in time length of a reboot or similar.

**Comment by iaguis** [comment]
LGTM

**Comment by jonboulle** [comment]
lgtm

**Pull request merged**